FirebaseStorageSwift/Tests/Integration/StorageAsyncAwait.swift

```diff
@@ -0,0 +1,391 @@
+      do {
+        _ = try await ref.putDataAsync(data)
+        XCTFail("Unexpected success from unauthorized putData")
+      } catch {
+        XCTAssertEqual((error as NSError).code, StorageErrorCode.unauthorized.rawValue)
+      }
+    }
+
+    func testSimplePutFile() throws {
+      let expectation = self.expectation(description: #function)
+      let putFileExpectation = self.expectation(description: "putFile")
+      let ref = storage.reference(withPath: "ios/public/testSimplePutFile")
+      let data = try XCTUnwrap("Hello Swift World".data(using: .utf8), "Data construction failed")
+      let tmpDirURL = URL(fileURLWithPath: NSTemporaryDirectory())
+      let fileURL = tmpDirURL.appendingPathComponent("hello.txt")
+      try data.write(to: fileURL, options: .atomicWrite)
+      let task = ref.putFile(from: fileURL) { result in
+        self.assertResultSuccess(result)
+        putFileExpectation.fulfill()
+      }
+
+      task.observe(StorageTaskStatus.success) { snapshot in
+        XCTAssertEqual(snapshot.description, "<State: Success>")
+        expectation.fulfill()
+      }
+
+      var uploadedBytes: Int64 = -1
+
+      task.observe(StorageTaskStatus.progress) { snapshot in
+        XCTAssertTrue(snapshot.description.starts(with: "<State: Progress") ||
+          snapshot.description.starts(with: "<State: Resume"))
+        guard let progress = snapshot.progress else {
+          XCTFail("Failed to get snapshot.progress")
+          return
+        }
+        XCTAssertGreaterThanOrEqual(progress.completedUnitCount, uploadedBytes)
+        uploadedBytes = progress.completedUnitCount
+      }
+      waitForExpectations()
+    }
+
+    func testAttemptToUploadDirectoryShouldFail() async throws {
+      // This `.numbers` file is actually a directory.
+      let fileName = "HomeImprovement.numbers"
+      let bundle = Bundle(for: StorageIntegrationCommon.self)
+      let fileURL = try XCTUnwrap(bundle.url(forResource: fileName, withExtension: ""),
+                                  "Failed to get filePath")
+      let ref = storage.reference(withPath: "ios/public/" + fileName)
+      do {
+        _ = try await ref.putFileAsync(from: fileURL)
+        XCTFail("Unexpected success from putFile of a directory")
+      } catch {
+        // TODO: Investigate generating a more descriptive error code than unknown.
+        let e = error as NSError
+        XCTAssertEqual(e.code, StorageErrorCode.unknown.rawValue)
+      }
+    }
+
+    func testPutFileWithSpecialCharacters() async throws {
+      let fileName = "hello&+@_ .txt"
+      let ref = storage.reference(withPath: "ios/public/" + fileName)
+      let data = try XCTUnwrap("Hello Swift World".data(using: .utf8), "Data construction failed")
+      let tmpDirURL = URL(fileURLWithPath: NSTemporaryDirectory())
+      let fileURL = tmpDirURL.appendingPathComponent("hello.txt")
+      try data.write(to: fileURL, options: .atomicWrite)
```

Comment:
optional: refactor into helper a method `func temporaryFile(named: String) -> URL` to use in this test and in `testSimplePutFileNoMetadata`

Reply:
I think this might not make the tests more readable.